src/signature.c

```diff
@@ -1340,3 +1391,161 @@ gboolean cms_verify_sig(GBytes *sig, X509_STORE *store, CMS_ContentInfo **cms, G
+	if (!incontent)
+		g_error("BIO_new_mem_buf() failed");
+
+	recipcerts = sk_X509_new_null();
+
+	/* load all recipient certificates from all provided PEM files */
+	for (gchar **recipcertpath = recipients; recipcertpath && *recipcertpath != NULL; recipcertpath++) {
+		STACK_OF(X509) *filecerts = load_certs_from_file(*recipcertpath, &ierror);
+		if (filecerts == NULL) {
+			g_propagate_error(error, ierror);
+			goto out;
+		}
+
+		/* add all recipient certs from recent file */
+		for (gint i = 0; i < sk_X509_num(filecerts); i++) {
+			sk_X509_push(recipcerts, sk_X509_value(filecerts, i));
+		}
+
+		sk_X509_free(filecerts);
+	}
+
+	cms = CMS_encrypt(recipcerts, incontent,
+			EVP_aes_256_cbc(), CMS_BINARY);
+	if (cms == NULL) {
+		unsigned long err;
+		const gchar *data;
+		int errflags;
+		err = ERR_get_error_line_data(NULL, NULL, &data, &errflags);
+		g_set_error(
+				error,
+				R_SIGNATURE_ERROR,
+				R_SIGNATURE_ERROR_INVALID,
+				"Failed to encrypt: %s", (errflags & ERR_TXT_STRING) ? data : ERR_error_string(err, NULL));
+		goto out;
+	}
+	if (!i2d_CMS_bio(outsig, cms)) {
+		g_set_error_literal(
+				error,
+				R_SIGNATURE_ERROR,
+				R_SIGNATURE_ERROR_SERIALIZE_SIG,
+				"Failed to serialize signature");
+		goto out;
+	}
+
+	res = bytes_from_bio(outsig);
+	if (!res) {
+		g_set_error_literal(
+				error,
+				R_SIGNATURE_ERROR,
+				R_SIGNATURE_ERROR_UNKNOWN,
+				"Read zero bytes");
+		goto out;
+	}
+
+	g_message("Encrypted for %d recipient%s", sk_X509_num(recipcerts), sk_X509_num(recipcerts) > 1 ? "s" : "");
+
+out:
+	ERR_print_errors_fp(stdout);
+	sk_X509_pop_free(recipcerts, X509_free);
+	BIO_free_all(incontent);
+	BIO_free_all(outsig);
+	return res;
+}
+
+GBytes *cms_decrypt(GBytes *content, const gchar *certfile, const gchar *keyfile, GError **error)
+{
+	GError *ierror = NULL;
+	g_autoptr(CMS_ContentInfo) icms = NULL;
+	g_autoptr(X509) decrypt_cert = NULL;
+	g_autoptr(EVP_PKEY) privkey = NULL;
+	gint decrypted;
+	BIO *outdecrypt = BIO_new(BIO_s_mem());
+	BIO *inenc = NULL;
+	GBytes *res = NULL;
+
+	g_return_val_if_fail(content != NULL, NULL);
+	g_return_val_if_fail(keyfile != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	inenc = BIO_new_mem_buf((void *)g_bytes_get_data(content, NULL),
+			g_bytes_get_size(content));
+	if (!inenc)
+		g_error("BIO_new_mem_buf() failed");
+
+	if (certfile) {
+		decrypt_cert = load_cert(certfile, &ierror);
+		if (decrypt_cert == NULL) {
+			g_propagate_error(error, ierror);
+			goto out;
+		}
+	}
+
+	privkey = load_key(keyfile, &ierror);
+	if (privkey == NULL) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	g_message("Decrypting signature...");
+
+	if (!(icms = d2i_CMS_bio(inenc, NULL))) {
+		g_set_error(
+				error,
+				R_SIGNATURE_ERROR,
+				R_SIGNATURE_ERROR_PARSE,
+				"Failed to parse CMS");
+		goto out;
+	}
+
+	/* assert we receied envelopedData */
+	g_assert_cmpint(OBJ_obj2nid(CMS_get0_type(icms)), ==, NID_pkcs7_enveloped);
+
```

Comment:
Shouldn't this be an a normal error case?

Reply:
As this depends on (probably broken) input data, I'd agree, yes. I make a GError with a nice error message out of it.